dynatrace/stackstate_checks/dynatrace/dynatrace.py

```diff
@@ -0,0 +1,259 @@
+        """
+        endpoint = self.url + "/api/v1/entity/infrastructure/process-groups"
+        process_groups = self.get_json_response(endpoint)
+        self.process_component(process_groups, "process-group")
+
+    def collect_services(self):
+        """
+        Collects all services from Dynatrace API and their relationships with other component types if exists
+        """
+        endpoint = self.url + "/api/v1/entity/services"
+        services = self.get_json_response(endpoint)
+        self.process_component(services, "service")
+
+    def process_component(self, response, component_type):
+        """
+        Process each component type and map those with specific data
+        :param response: Response of each component type endpoint
+        :param component_type: Component type
+        :return: create the component on stackstate API
+        """
+        urn = ''
+        if "error" not in response:
+            for item in response:
+                # special case for host type as we get some float values
+                if component_type == "host":
+                    for key in item.keys():
+                        if type(item[key]) is float:
+                            item[key] = int(item[key])
+                identifiers = []
+                labels = []
+                externalId = item.get("entityId")
+                if component_type == "service":
+                    urn = "urn:service:/{}".format(externalId)
+                elif component_type == "process-group":
+                    urn = "urn:process-group:/{}".format(externalId)
+                elif component_type == "application":
+                    urn = "urn:application:/{}".format(externalId)
+                elif component_type == "process":
+                    urn = "urn:process:/{}".format(externalId)
+                elif component_type == "host":
+                    displayName = item.get("displayName")
+                    urn = "urn:host:/{}".format(displayName)
+                identifiers.append(urn)
+                tags = self.add_tags(item) + self.tags
+                if "tags" in item:
+                    del item["tags"]
+                labels = self.add_labels(item, labels)
+                data = {
+                    "identifiers": identifiers,
+                    "tags": tags,
+                    "domain": self.domain,
+                    "environment": self.environment,
+                    "instance": self.url,
+                    "labels": labels
+                }
+                data.update(item)
+                data = self.filter_data(data)
+                self.component(externalId, component_type, data)
+                self.collect_relations(item, externalId)
+        else:
+            self.log.info("Problem getting the {0} or No {1} found.".format(component_type, component_type))
+            self.log.error("Response code {0} with message : {1}".format(response["error"].get("code"),
```

Comment:
1. When there is an error we shouldn't just log and continue, because sending a topology snapshot without the type means the topology will be completely clear of that type. In the case of an error the whole check run should be aborted. I would thus raise a specific home made exception, add as much info as possible and let the general purpose exception handler in the `check` method take care of the rest. 
2. This error handler does not belong here. This function should simply be able to assume the response is a successful one.

Reply:
done.